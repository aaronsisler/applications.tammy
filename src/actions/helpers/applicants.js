export const setApplicants = (applicants) => ({
    type: 'SET_APPLICANTS',
    applicants
});
