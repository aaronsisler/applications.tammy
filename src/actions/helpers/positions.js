export const setPositions = (positions) => ({
    type: 'SET_POSITIONS',
    positions
});
