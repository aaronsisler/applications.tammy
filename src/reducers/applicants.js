const applicantsReducerDefaultState = [];

const applicantsReducer = (state = applicantsReducerDefaultState, action) => {
    switch (action.type) {
        case 'SET_APPLICANTS':
            return action.applicants || applicantsReducerDefaultState;
        default:
            return state;
    }
};

export default applicantsReducer;

