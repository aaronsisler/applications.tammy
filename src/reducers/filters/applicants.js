const applicantsFilterReducerDefaultState =
{
    text: '',
};

const applicantsFilterReducer = (state = applicantsFilterReducerDefaultState, action) => {
    switch (action.type) {
        case 'SET_APPLICANTS_TEXT_FILTER':
            return {
                ...state,
                text: action.text
            };
        default:
            return state;
    }
};

export default applicantsFilterReducer;
