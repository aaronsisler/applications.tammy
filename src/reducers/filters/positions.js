const positionsFilterReducerDefaultState =
{
    text: '',
};

const positionsFilterReducer = (state = positionsFilterReducerDefaultState, action) => {
    switch (action.type) {
        case 'SET_POSITIONS_TEXT_FILTER':
            return {
                ...state,
                text: action.text
            };
        case 'CLEAR_POSITIONS_TEXT_FILTER':
            return {
                ...state,
                text: ''
            };
        default:
            return state;
    }
};

export default positionsFilterReducer;
