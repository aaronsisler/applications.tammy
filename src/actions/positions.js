import database from '../firebase/firebase';
import { setPositions } from 'Actions/helpers/positions';

export const startSetPositions = () => (dispatch) =>
    database.ref(`positions`).once('value').then((snapshot) => {
        const positions = [];
        snapshot.forEach((childSnapshot) => {
            positions.push({
                positionId: childSnapshot.key,
                ...childSnapshot.val(),
            });
        });

        dispatch(setPositions(positions));
    });
