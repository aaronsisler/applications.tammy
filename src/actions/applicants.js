import database from 'Firebase/firebase';
import { setApplicants } from 'Actions/helpers/applicants';
import { setApplicantsTextFilter } from 'Actions/filters/applicantsText';

export const startSetApplicants = () => (dispatch, getState) => {
    const { positionId } = getState().workflow.position;

    return database.ref(`applicants/${positionId}`).once('value').then((snapshot) => {
        const applicants = [];
        snapshot.forEach((childSnapshot) => {
            const applicant = childSnapshot.val();
            const applicantNotes = applicant.applicantNotes
                ? [...Object.values(applicant.applicantNotes)]
                : [];
            applicants.push({
                applicantId: childSnapshot.key,
                positionId,
                ...childSnapshot.val(),
                applicantNotes,
            });
        });

        return dispatch(setApplicants(applicants));
    });
}

export const startSetApplicantsTextFilter = (text) => (dispatch) => dispatch(setApplicantsTextFilter(text));
