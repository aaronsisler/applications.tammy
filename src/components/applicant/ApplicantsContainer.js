import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { startSetApplicants } from 'Actions/applicants';
import selectApplicants from 'Selectors/applicants';
import ApplicantDetails from 'Applicant/ApplicantDetails';
import ApplicantsList from 'Applicant/ApplicantsList';
import ApplicantsListFilter from 'Applicant/ApplicantsListFilter';

export class ApplicantsContainer extends React.Component {
    constructor(props) {
        super(props);
        // props.startSetApplicants();
    }

    render() {
        return (
            <div className="inbox_container">
                {this.props.applicants &&
                    <div className="inbox_widget">
                        <div className="inbox_list">
                            <ApplicantsListFilter />
                            <ApplicantsList applicants={this.props.applicants} />
                        </div>
                        <ApplicantDetails />
                    </div>
                }
            </div>
        );
    }
}

/* istanbul ignore next */
const mapStateToProps = (state) => ({
    applicants: selectApplicants(state.applicants, state.filters.applicants),
});

/* istanbul ignore next */
const mapDispatchToProps = (dispatch) => ({
    startSetApplicants: () => dispatch(startSetApplicants()),
});

export default connect(mapStateToProps, mapDispatchToProps)(ApplicantsContainer);

ApplicantsContainer.propTypes = {
    applicants: PropTypes.array,
    startSetApplicants: PropTypes.func.isRequired,
};
