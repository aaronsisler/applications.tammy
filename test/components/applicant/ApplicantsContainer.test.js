import React from 'react';
import { shallow } from 'enzyme';
import { ApplicantsContainer } from 'Applicant/ApplicantsContainer';
import applicants, { applicantId } from '../../fixtures/applicants';
import { position, positionId } from '../../fixtures/positions';
import history from 'Tools/history';
jest.mock('Tools/history');

describe('ApplicantsContainer', () => {
    let wrapper;
    const push = jest.fn();
    const startClearApplicants = jest.fn();
    const startSetApplicants = jest.fn();

    const buildWrapper = (positionIdInput, positionInput, applicantIdInput) => {
        wrapper = shallow(
            <ApplicantsContainer
                applicantId={applicantIdInput}
                applicants={applicants}
                positionId={positionIdInput}
                position={positionInput}
                startClearApplicants={startClearApplicants}
                startSetApplicants={startSetApplicants}
            />
        );
    };

    it('should call startClearApplicants on unmount', () => {
        buildWrapper(positionId, position, undefined);

        wrapper.unmount();

        expect(startClearApplicants).toHaveBeenCalled();
    });

    describe('when position is found', () => {
        it('should call startSetApplicants on mount', () => {
            buildWrapper(positionId, position, undefined);

            expect(startSetApplicants).toHaveBeenLastCalledWith(positionId);
        });

        describe('when applicantId is provided', () => {
            it('should render ApplicantsContainer correctly', () => {
                buildWrapper(positionId, position, applicantId);

                expect(wrapper).toMatchSnapshot();
            });
        });

        describe('when applicantId is NOT provided', () => {
            it('should render ApplicantsContainer correctly', () => {
                buildWrapper(positionId, position, undefined);

                expect(wrapper).toMatchSnapshot();
            });
        });
    });

    describe('when position is NOT found', () => {
        it('should call history.push with correct route', () => {
            history.push = push;
            buildWrapper(positionId, undefined, undefined);

            expect(push).toHaveBeenLastCalledWith(`/not_found`);
        })
    });
});
