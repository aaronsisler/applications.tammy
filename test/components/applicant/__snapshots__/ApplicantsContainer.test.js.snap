// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApplicantsContainer should render ApplicantsContainer correctly 1`] = `
<div
  className="inbox_container"
>
  <div
    className="inbox_widget"
  >
    <div
      className="inbox_list"
    >
      <ApplicantsList
        applicants={
          Array [
            Object {
              "applicantId": "firstApplicantId",
              "applicantNotes": Array [
                Object {
                  "statusNote": "First note",
                },
                Object {
                  "statusNote": "Second note",
                },
              ],
              "positionId": "firstPositionId",
              "user": Object {
                "addressLine1": "123 Evergreen Terrace",
                "addressLine2": "Apt. 456",
                "city": "Durham",
                "displayName": "Johnny Appleseed",
                "email": "johnny.appleseed@gmail.com",
                "firstName": "Johnny",
                "lastName": "Appleseed",
                "middleName": "Middleton",
                "phoneAreaCode": "919",
                "phoneExt": "12345",
                "phoneLineNumber": "4567",
                "phonePrefix": "123",
                "postalCode": "27540",
                "state": "NC",
              },
            },
            Object {
              "applicantId": "secondApplicantId",
              "applicantNotes": Array [
                Object {
                  "statusNote": "First note",
                },
                Object {
                  "statusNote": "Second note",
                },
              ],
              "positionId": "firstPositionId",
              "user": Object {
                "addressLine1": "123 Evergreen Terrace",
                "addressLine2": "Apt. 456",
                "city": "Durham",
                "displayName": "Johnny Appleseed",
                "email": "johnny.appleseed@gmail.com",
                "firstName": "Johnny",
                "lastName": "Appleseed",
                "middleName": "Middleton",
                "phoneAreaCode": "919",
                "phoneExt": "12345",
                "phoneLineNumber": "4567",
                "phonePrefix": "123",
                "postalCode": "27540",
                "state": "NC",
              },
            },
          ]
        }
      />
    </div>
    <Connect(ApplicantDetails) />
  </div>
</div>
`;
